lib/pin_depends.ml

```diff
@@ -106,3 +108,124 @@ let get_opam ~job ~pkg url =
+      (function
+        | Tar.Header.End_of_stream -> Lwt.return opam_file'
+        | exn -> Lwt.fail exn)
+  in
+  Lwt_unix.openfile (Fpath.to_string tarball) [ Unix.O_RDONLY ] 0 >>= fun fd ->
+  Lwt.finalize
+    (fun () ->
+      let tgz = Tgz.of_in_channel ~internal:(Cstruct.create 4096) fd in
+      find None tgz)
+    (fun () -> Lwt_unix.close fd)
+
+let fetch ~transport ~path ~tarball =
+  let open Cohttp in
+  let open Cohttp_lwt_unix in
+  Client.get (transport ^ "://" ^ path |> Uri.of_string) >>= fun (resp, _) ->
+  if resp |> Response.status |> Code.code_of_status <> 302 then
+    Fmt.failwith "Expected redirect response.";
+  let location =
+    resp |> Response.headers |> Fun.flip Header.get "Location" |> Option.get
+  in
+  Client.get (location |> Uri.of_string) >>= fun (resp, body) ->
+  if resp |> Response.status |> Code.code_of_status <> 200 then
+    Fmt.failwith "Failed to download the pin-depend: %a" Response.pp_hum resp;
+  Lwt_io.with_file ~mode:Lwt_io.Output (Fpath.to_string tarball) @@ fun ch ->
+  Lwt_stream.iter_s (Lwt_io.write ch) (Cohttp_lwt.Body.to_stream body)
+
+let get_opam_http ~job ~pkg url =
+  ignore job;
+  (* TODO: support cancellation *)
+  let { OpamUrl.transport; path; hash = _; backend = _ } = url in
+  match Astring.String.cuts ~sep:"/" path with
+  | [ "github.com"; _org; repo; "archive"; hash; archive ] -> (
+      if not (Str.string_match re_hash hash 0) then
+        Fmt.failwith "Invalid commit hash %S for package %s" hash
+          (OpamPackage.to_string pkg);
+      if
+        not
+          Astring.String.(
+            is_suffix ~affix:".tar.gz" archive
+            || is_suffix ~affix:".tgz" archive)
+      then
+        Fmt.failwith "Only gzipped tar archives are supported (for %s)" archive;
+      let repo_url = OpamUrl.base_url url in
+      let tarball = Cmd.local_copy repo_url in
+
+      (if Cmd.file_exists tarball then Lwt.return_unit
+       else
+         Lwt.catch
+           (fun () -> fetch ~transport ~path ~tarball)
+           (fun exn ->
+             Lwt_unix.unlink (Fpath.to_string tarball) >>= fun () ->
+             Lwt.fail exn))
+      >>= fun () ->
+      read_opam_file ~tarball ~prefix:(repo ^ "-" ^ hash) pkg >>= function
+      | Some contents -> Lwt.return contents
+      | None ->
+          Fmt.failwith "Couldn't get opam file %a contents from %s for %s"
+            Fpath.pp tarball (OpamUrl.to_string url)
+            (OpamPackage.to_string pkg))
+  | _ ->
+      Fmt.failwith
+        "Only Git or GitHub https pin-depends are supported (got %S for %s)"
+        (OpamUrl.to_string url)
+        (OpamPackage.to_string pkg)
+
+let get_opam ~job ~pkg url =
+  match url.OpamUrl.backend with
+  | `git -> get_opam_git ~job ~pkg url
+  | `http -> get_opam_http ~job ~pkg url
+  | _ ->
+      Fmt.failwith
+        "Only Git or GitHub https pin-depends are supported (got %S for %s)"
+        (OpamUrl.to_string url)
+        (OpamPackage.to_string pkg)
```

Comment:
Possible to add a test around this functionality?

Reply:
Well if we switch ocaml-ci to pin-depend on the tailwindcss tarball, ocaml-ci's own deployment will test the feature!